[Storefront Preview] SCAPI cache breaker for older v2 PWA Kit (@W-15214446@)

vmarta opened the pull request:
Porting the changes from #1701. Unfortunately, while the solution/concept is still the same, the implementation had to be different. 

In the older V2 PWA Kit:
- The `StorefrontPreview` component does not have access to the commerce api clients because they're not in the same package
- The `Proxy` mechanism has already been used. So we need to add the cache breaker using the existing mechanism (see `willSendRequest`)

Todos:
- [x] Write tests
- [ ] Write some documentation for existing projects to consume the update?
- [x] Bump up version number? To release as a patch.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

https://runtime.commercecloud.com/mobify/blitz/will/preview?path=/global/en-GB/product/25695327M?color%3DJJI15XX%26size%3D9SM%26pid%3D701644153083M

1. Note that the product price should be the regular price
2. In the Preview form, enter tester=yes as the custom qualifier
3. Click on the Preview button
4. The price is immediately updated with the discounted promo price.

Note: to verify that the environment is actually using a bundle for the _older_ PWA Kit, you can add an item to the cart, and that should result in an OCAPI call (not SCAPI).

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
Update commerce-sdk-isomorphic dependency

And updated all the package-lock.json files too.

vmarta committed:
Break SCAPI cache with a different way for v2 PWA Kit

vmarta committed:
No longer import from commerce-api

And provided a default value for `apiClients`.

vmarta committed:
Simplify logic

vmarta committed:
Increase the build/main.js max size slightly

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/components/storefront-preview/index.jsx:
```diff
@@ -46,6 +48,26 @@ export const StorefrontPreview = ({
+
+    useEffect(() => {
+        if (enabled && isHostTrusted) {
+            const originalWillSendRequest = apiClients.willSendRequest
```
FYI the existing mechanism has this `willSendRequest`, which will get called before requesting SCAPI: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/eab57e523f2a8b2bf961d24a51a7fbe9f8310591/packages/template-retail-react-app/app/commerce-api/index.js#L147-L151

By default, `willSendRequest` will automatically inject the access token into the request header.

vmarta commented on packages/template-retail-react-app/app/components/_app/index.jsx:
```diff
@@ -153,7 +155,10 @@ const App = (props) => {
             {/* For pwa-kit v2, we need to use this getToken prop to make sure it can work properly
                 with Storefront Preview feature*/}
-            <StorefrontPreview getToken={() => window.localStorage.getItem('token')}>
+            <StorefrontPreview
+                getToken={() => window.localStorage.getItem('token')}
+                apiClients={apiClients}
+            >
```
To consume the update, existing projects will need to do the following:
- Set a new `apiClients` prop for the `StorefrontPreview` component
- Update their dependencies for: `pwa-kit-react-sdk` and `commerce-sdk-isomorphic`

bendvc replied:
@vmarta if you had to give me a jsdoc style description of that the `apiClient` props is, what would it be?

vmarta replied:
@bendvc there's already a jsdoc comment describing that `apiClient` prop (see the other file "packages/pwa-kit-react-sdk/src/ssr/universal/components/storefront-preview/index.jsx"). Did you mean that that is not descriptive enough?
`* @param {Object} apiClients - CommerceAPI instance`

bendvc replied:
Yes... the comment doesn't really let the end user know what it does or why you can pass the api clients to it. 

The reason I'm asking this question is because I think it might be hard to describe why we are passing that object in in 1 or 2 sentences which might be a code smell that you are doing something wrong. 

Looking at the pattern you have, it doesn't really follow typical patterns for components. Usually you pass in a prop and it either will be used as options or a conditional or rendering, or used directly to display data that you are passing in. 

The pattern you have here is you pass in an object to the component and when that component mounts you modify the object, and that prop isn't used again.

I think there is probably a better pattern that we can create that won't require use to add this apiClients props onto the component, but maybe something more generic.

vmarta commented on packages/template-retail-react-app/package.json:
```diff
@@ -26,7 +26,7 @@
         "bundlesize2": "^0.0.31",
         "card-validator": "^8.1.1",
-        "commerce-sdk-isomorphic": "^1.9.0",
+        "commerce-sdk-isomorphic": "^1.12.0",
```
Needs to use a newer version of commerce-sdk-isomorphic that supports custom parameter (for us to pass in our own `c_cache_breaker`).

adamraya commented on packages/pwa-kit-react-sdk/src/ssr/universal/components/storefront-preview/index.jsx:
The approach of re-using `willSendRequest` looks good but I tested the change and I'm not seeing the cache breaker working.

@vmarta Do you have a v2 bundle pushed to Runtime Admin to test the change?

Do we also want to add a test to confirm the cache breaker is working as expected?

vmarta replied:
@adamraya I don't have a v2 bundle yet for testing. I'll make one.

And for the tests, yes I'm writing them now.

vmarta replied:
Ok, here's the v2 bundle (and I've updated the PR description too with this instruction):

https://runtime.commercecloud.com/mobify/blitz/will/preview?path=/global/en-GB/product/25695327M?color%3DJJI15XX%26size%3D9SM%26pid%3D701644153083M

1. Note that the product price should be the regular price
2. In the Preview form, enter tester=yes as the custom qualifier
3. Click on the Preview button
4. The price is immediately updated with the discounted promo price.

Note: to verify that the environment is actually using a bundle for the _older_ PWA Kit, you can add an item to the cart, and that should result in an OCAPI call (not SCAPI).

vmarta committed:
Add tests

vmarta committed:
Refactor to remove the tight coupling

Now StorefrontPreview component no longer knows about the inner workings of the api clients.

vmarta committed:
Update jsdoc

vmarta committed:
Also refactor the tests

vmarta committed:
Restore console.log mock

vmarta committed:
Fix tests

vmarta committed:
Bump up patch version

vmarta committed:
Update changelog

vmarta committed:
Update package lock files

vmarta commented on packages/template-retail-react-app/app/components/_app/index.jsx:
```diff
@@ -153,7 +172,10 @@ const App = (props) => {
             {/* For pwa-kit v2, we need to use this getToken prop to make sure it can work properly
                 with Storefront Preview feature*/}
-            <StorefrontPreview getToken={() => window.localStorage.getItem('token')}>
+            <StorefrontPreview
+                getToken={() => window.localStorage.getItem('token')}
+                onInit={initStorefrontPreview}
+            >
```
Now removing the coupling between the template and pwa-kit-react-sdk:
By passing in a callback, then it would be the template that defines the wiring of `apiClients.willSendRequest` -> this is better than what I had, where StorefrontPreview somehow knew about the inner workings of the api clients.

vmarta commented:
Opening a new PR #1740
